crates/plugins/src/manager.rs

```diff
@@ -0,0 +1,219 @@
+    pub fn uninstall(&self, plugin_name: &str) -> Result<bool> {
+        let plugin_store = self.store();
+        let manifest_file = plugin_store.installed_manifest_path(plugin_name);
+        match manifest_file.exists() {
+            // Remove the manifest and the plugin installation directory
+            true => {
+                fs::remove_file(manifest_file)?;
+                fs::remove_dir_all(plugin_store.plugin_subdirectory_path(plugin_name))?;
+                Ok(true)
+            }
+            false => Ok(false),
+        }
+    }
+
+    /// Checks manifest to see if the plugin is compatible with the running version of Spin, does
+    /// not have a conflicting name with Spin internal commands, and is not a downgrade of a
+    /// currently installed plugin.
+    pub fn check_manifest(
+        &self,
+        plugin_manifest: &PluginManifest,
+        spin_version: &str,
+        allow_downgrades: bool,
+    ) -> Result<()> {
+        // Disallow installing plugins with the same name as spin internal subcommands
+        if SPIN_INTERNAL_COMMANDS
+            .iter()
+            .any(|&s| s == plugin_manifest.name())
+        {
+            bail!(
+                "Can't install a plugin with the same name ({}) as an internal command",
+                plugin_manifest.name()
+            );
+        }
+
+        // Disallow reinstalling identical plugins and downgrading unless permitted.
+        if let Ok(installed) = self.store.read_plugin_manifest(&plugin_manifest.name()) {
+            if &installed == plugin_manifest {
+                bail!(
+                    "Plugin {} is already installed with version {}.",
+                    plugin_manifest.name(),
+                    installed.version,
+                );
+            } else if installed.version > plugin_manifest.version && !allow_downgrades {
+                bail!(
+                    "Newer version {} of {} plugin is already installed. To downgrade to version {} set the `--downgrade` flag.",
+                    installed.version,
+                    plugin_manifest.name(),
+                    plugin_manifest.version,
+                );
+            }
+        }
+
+        check_supported_version(plugin_manifest, spin_version)
+    }
+
+    /// Fetches a manifest from a local, remote, or repository location and returned the parsed
+    /// PluginManifest object.
+    pub async fn get_manifest(
+        &self,
+        manifest_location: &ManifestLocation,
+    ) -> Result<PluginManifest> {
+        let plugin_manifest = match manifest_location {
+            ManifestLocation::Remote(url) => {
+                log::info!("Pulling manifest for plugin from {}", url);
+                reqwest::get(url.as_ref())
+                    .await?
+                    .error_for_status()
+                    .with_context(|| format!("Manifest endpoint {} is unhealthy", url))?
```

Comment:
"unhealthy" may be a little confusing here; it really just means the request failed, right?

Reply:
Good point. This now maps to a custom `Error::ConnectionFailed`.